### PR TITLE
[videolecture] add support for sites using viidea service and add support for multi part videos (fixes #7384)

### LINKED ISSUE #7384: Not able to download several videos/parts from the same videolectures.net url
Hi,

First of all thank you for this wonderful tool. I was trying to download the videos from a videolectures.net url (youtube-dl http://videolectures.net/mlss09uk_bishop_ibi/) but I have not been able to download all the videos that are listed in the web page (part1 and part2), it always downloads the first part. Is it possible to select which part to download? (I have used the playlist options but they seem not to work in this case).  I have used youtube-dl version 2015.11.02.

### PR DESCRIPTION
**Pull request by remitamine**

### CODE REVIEW

**Comment by dstftw** [comment]
```
F
======================================================================
FAIL: test_no_duplicates (__main__.TestAllURLsMatching)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/test_all_urls.py", line 97, in test_no_duplicates
    '%s should not match URL %r . That URL belongs to %s.' % (type(ie).__name__, url, tc['name']))
AssertionError: ViideaIE should not match URL u'http://videolectures.net/promogram_igor_mekjavic_eng/video/1/smil.xml' . That URL belongs to Generic.

----------------------------------------------------------------------
Ran 1 test in 5.469s

FAILED (failures=1)
```

**Comment by remitamine** [comment]
i will add `(?!/\d+.xml)` to the end of regex.

**Comment by dstftw** [comment]
No, add the same suffix as videolectures extractor has now in master.

**Comment by remitamine** [comment]
ok i will add it.

**Comment by remitamine** [comment]
i added the suffix from master branch and fixed the multipart extraction.

File: youtube_dl/extractor/viidea.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
+from __future__ import unicode_literals
+
+import re
+
+from .common import InfoExtractor
+from ..compat import (
+    compat_HTTPError,
```
Comment (dstftw) [inline youtube_dl/extractor/viidea.py]: This is unused.

File: youtube_dl/extractor/viidea.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    compat_urlparse,
+)
+from ..utils import (
+    ExtractorError,
```
Comment (dstftw) [inline youtube_dl/extractor/viidea.py]: And this as well.

File: youtube_dl/extractor/viidea.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        }],
+        'playlist_count': 2,
+    }]
+
+    def _real_extract(self, url):
+        lecture_slug, part = re.match(self._VALID_URL, url).groups()
+
+        webpage = self._download_webpage(url, lecture_slug)
+
+        cfg = self._parse_json(self._search_regex(r'cfg\s*:\s*({[^}]+})', webpage, 'cfg'), lecture_slug, js_to_json)
```
Comment (dstftw) [inline youtube_dl/extractor/viidea.py]: This will fail if `cfg` turns out to contain inner dictionary.

File: youtube_dl/extractor/viidea.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+        lecture_id = str(cfg['obj_id'])
```
Comment (dstftw) [inline youtube_dl/extractor/viidea.py]: Use `compat_str`.

File: youtube_dl/extractor/viidea.py (hunk 3 above)
Comment (remitamine) [inline youtube_dl/extractor/viidea.py]: and this `cfg\s*:\s*({.+?}),xhr`?

Comment (dstftw) [inline youtube_dl/extractor/viidea.py]: I also saw `cfg:{.+?},ga:function(`, so let's just add both:
```python
[r'cfg\s*:\s*({.+?}),[\da-zA-Z_]:\(?function', r'cfg\s*:\s*({[^}]+})']
```

File: youtube_dl/extractor/viidea.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                info['id'] = '%s_part%s' % (lecture_id, part)
+                switch = smil.find('.//switch')
+                if switch is not None:
+                    info['duration'] = parse_duration(switch.attrib.get('dur'))
+                return info
+            else:
+                for part in parts:
+                    entries.append(self.url_result('%s/%s/video/%s' % (base_url, lecture_slug, part), 'Viidea'))
+                lecture_info['_type'] = 'multi_video'
+        else:
```
Comment (dstftw) [inline youtube_dl/extractor/viidea.py]: This branch should probably take place always when `cfg['type']` equals `"evt"` regardless of the `cfg['videos']` content - potentially `videos` of `evt` item may contain some promo videos that will "shadow" the actual videos of this event with current approach.

Comment (remitamine) [inline youtube_dl/extractor/viidea.py]: may be there are other types that contain playlists?

Comment (dstftw) [inline youtube_dl/extractor/viidea.py]: Maybe, who knows. At least we can try playlist if videos is empty and if type is evt (or not vpr).

Comment (remitamine) [inline youtube_dl/extractor/viidea.py]: ok i will add the condition.

**Comment by remitamine** [comment]
i made the requested changes.

File: youtube_dl/extractor/viidea.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            'display_id': lecture_slug,
+            'title': lecture_data['title'],
+            'timestamp': parse_iso8601(lecture_data.get('time')),
+            'description': lecture_data.get('description_wiki'),
+            'thumbnail': lecture_data.get('thumb'),
+        }
+
+        entries = []
+        parts = cfg.get('videos')
+        if lecture_data.get('type') != 'evt' and parts:
```
Comment (dstftw) [inline youtube_dl/extractor/viidea.py]: Actually I mean to process `videos` always when `videos` present and also try playlist when either no `videos` present or `type` is `evt`.

**Comment by remitamine** [comment]
is this what you mean(e7a90483509634939824f408e25ccc7da6cef350)

**Comment by dstftw** [comment]
Yes.

File: youtube_dl/extractor/viidea.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    }]
+
+    def _real_extract(self, url):
+        lecture_slug, part = re.match(self._VALID_URL, url).groups()
+
+        webpage = self._download_webpage(url, lecture_slug)
+
+        cfg = self._parse_json(self._search_regex([r'cfg\s*:\s*({.+?}),[\da-zA-Z_]:\(?function', r'cfg\s*:\s*({[^}]+})'], webpage, 'cfg'), lecture_slug, js_to_json)
+
+        lecture_id = str(cfg['obj_id'])
```
Comment (dstftw) [inline youtube_dl/extractor/viidea.py]: One more to `compat_str`.

**Comment by remitamine** [comment]
i changed it to compat_str(87e59de943eef05b8bedb8878699b50bd8556560)

**Comment by dstftw** [comment]
Cherry-picked. Thanks.

**Pull request closed without merging**